service_capacity_modeling/models/common.py

```diff
@@ -689,3 +692,198 @@ def merge_plan(
     return CapacityPlan(
         requirements=merged_requirements, candidate_clusters=merged_clusters
     )
+
+
+def derived_buffer_for_component(buffer: Dict[str, Buffer], components: List[str]):
+    scale = 0.0
+    preserve = False
+
+    if not buffer:
+        return scale, preserve
+
+    for bfr in buffer.values():
+        if any(component in components for component in bfr.components):
+            if bfr.intent == BufferIntent.scale:
+                scale = max(scale, bfr.ratio)
+            if bfr.intent == BufferIntent.preserve:
+                preserve = True
+
+    return scale, preserve
+
+
+def get_cores_from_current_capacity(
+    current_capacity: CurrentClusterCapacity, buffers: Buffers, instance: Instance
+):
+    # compute cores required per zone
+    cpu_success_buffer = (1 - cpu_headroom_target(instance, buffers)) * 100
+    current_cpu_utilization = current_capacity.cpu_utilization.mid
+
+    if current_capacity.cluster_instance is None:
+        cluster_instance = shapes.instance(current_capacity.cluster_instance_name)
+    else:
+        cluster_instance = current_capacity.cluster_instance
+
+    current_cores = cluster_instance.cpu * current_capacity.cluster_instance_count.mid
+
+    scale, preserve = derived_buffer_for_component(buffers.derived, ["compute", "cpu"])
+    # Scale and preserve for the same component should not be passed together.
+    # If user passes it, then scale will be preferred over preserve.
+    if scale > 0:
```

Comment:
shouldn't scale be > 1 here? If scale is 0.5, then we effectively want to preserve?

Reply:
Here, `core_scale_up_factor = max(1.0, new_cpu_utilization / cpu_success_buffer)` we are taking the max of 1 or  new cpu factor. If new CPU factor is low then we will keep the existing cores.